textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -0,0 +1,167 @@
+package edu.uci.ics.textdb.dataflow.regexmatch;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import edu.uci.ics.textdb.common.constants.DataConstants;
+
+
+public class GramBooleanQuery {
+	public enum QueryOp {
+		NONE, // doesn't match any string
+		ANY,  // matches any string
+		
+		AND,
+		OR
+	}
+	QueryOp operator;
+	List<String> operandList;
+	List<GramBooleanQuery> subQueryList;
+	
+	private static int gramNum = 3;
+	
+
+	public GramBooleanQuery(QueryOp operator) {
+		this.operator = operator;
+		operandList = new ArrayList<String>();
+		subQueryList = new ArrayList<GramBooleanQuery>();
+	}
+	
+	public boolean equals(GramBooleanQuery query) {
+		if (this.operator != query.operator
+			|| this.operandList.size() != query.operandList.size()
+			|| this.subQueryList.size() != query.subQueryList.size()) {
+			return false;
+		}
+		
+		Set<String> operandSet = new HashSet<String>(this.operandList);
+		if (!operandSet.equals(new HashSet<String>(query.operandList))) {
+			return false;
+		}
+		
+		if (this.subQueryList.size() == 0) {
+			return true;
+		}
+		
+		int[] used = new int[this.subQueryList.size()];
+		return this.equalsHelper(query, used, 0);
+	}
+	/**
+	 * This is a helper function called by {@code equals} function.
+	 * It takes a DFS approach to recursively determine 
+	 * whether two {@code TrigramBooleanQuery} list contains same set of elements. 
+	 * @param query
+	 * @param isUsed
+	 * @param index
+	 * @return
+	 */
+	private boolean equalsHelper(GramBooleanQuery query, int[] isUsed, int index) {
+		if (index == query.subQueryList.size()) {
+			return true;
+		}
+		
+		for (int i = 0; i < query.subQueryList.size(); i++) {
+			if (isUsed[i] == 1) continue;
+			if (this.subQueryList.get(index).equals(query.subQueryList.get(i))) {
+				isUsed[i] = 1;
+				if (equalsHelper(query, isUsed, index+1)){
+					return true;
+				} else {
+					isUsed[i] = 0;
+				}
+			}
+		}
+		return false;
+	}
+	
+	public void add(ArrayList<String> list) {
```

Comment:
this add function takes a list of strings,
for example, {abcd, wxyz}, then:
trigrams({abcd, wxyz}) = trigrams(abcd) OR trigrams(wxyz) = (abc AND bcd) OR (wxy AND xyz)

So for a list of strings, we assume OR.
For a single string we assume AND when breaking it into n-grams.